lib/request_validations/conversions.ex

```diff
@@ -0,0 +1,147 @@
+    
+    ## Examples
+    
+        iex> cast(%{name: "Butterfree", number: "012"}, [name: :string, number: :integer])
+        {:ok, %{name: "Butterfree", number: 12}}
+  """
+  @spec cast(map(), Keyword.t()) :: {:ok, map()} | {:error, [ConversionError.t()]}
+  def cast(attrs, fields), do: cast(attrs, fields, [])
+
+  defp cast(attrs, [], []), do: {:ok, attrs}
+  defp cast(_attrs, [], errors), do: {:error, errors}
+
+  defp cast(attrs, [{key, type} | rest], errors) do
+    case cast_one(attrs, key, type) do
+      {:ok, new_attrs} -> cast(new_attrs, rest, errors)
+      {:error, error} -> cast(attrs, rest, [error | errors])
+    end
+  end
+
+  defp cast_one(attrs, key, type) do
+    case Map.fetch(attrs, key) do
+      {:ok, value} -> cast_one(attrs, key, value, type)
+      :error -> {:ok, attrs}
+    end
+  end
+
+  defp cast_one(attrs, key, value, type) do
+    case parse(value, type) do
+      {:ok, result} -> {:ok, Map.put(attrs, key, result)}
+      {:error, _reason} -> {:error, convertion_error(key, type)}
+    end
+  end
+
+  @spec parse(String.t(), :date) :: {:ok, Date.t()} | {:error, atom()}
+  @spec parse(String.t(), :integer) :: {:ok, integer()} | {:error, :bad_integer}
+  @spec parse(String.t(), :boolean) :: {:ok, boolean()} | {:error, :bad_boolean}
+  @spec parse(String.t() | [String.t()], :list) :: {:ok, [String.t()]}
+  @spec parse(String.t(), :string) :: {:ok, String.t()}
+  def parse(value, :date), do: Date.from_iso8601(value)
+  def parse(value, :integer), do: parse_integer(value)
+  def parse(value, :float), do: parse_float(value)
+  def parse(value, :boolean), do: parse_boolean(value)
+  def parse(value, :list), do: parse_list(value)
+  def parse(value, :string), do: parse_string(value)
+  def parse(value, :uuid), do: parse_uuid(value)
+
+  @spec parse_string(term()) :: {:ok, String.t()} | {:error, :bad_string}
+  def parse_string(value) when is_binary(value), do: {:ok, value}
+  def parse_string(_value), do: {:error, :bad_string}
+  @spec parse_integer(term()) :: {:ok, integer()} | {:error, :bad_integer}
+  def parse_integer(value) do
+    case Integer.parse(value) do
+      {value, ""} -> {:ok, value}
+      _error -> {:error, :bad_integer}
+    end
+  end
+
+  @spec parse_float(term()) :: {:ok, float()} | {:error, :bad_float}
+  def parse_float(value) do
+    case Float.parse(value) do
+      {value, ""} -> {:ok, value}
+      _error -> {:error, :bad_float}
+    end
+  end
+
+  @spec parse_boolean(term()) :: {:ok, boolean()} | {:error, :bad_boolean}
+  def parse_boolean("true"), do: {:ok, true}
+  def parse_boolean("false"), do: {:ok, false}
+  def parse_boolean(_unexpeted), do: {:error, :bad_boolean}
+
+  @spec parse_list(String.t() | [String.t()]) :: {:ok, [String.t()]} | {:error, :bad_list}
+  def parse_list(list) when is_list(list), do: {:ok, list}
+  def parse_list(val) when is_binary(val), do: {:ok, String.split(val, ",")}
+  def parse_list(_unexpected), do: {:error, :bad_list}
+
+  def parse_uuid(uuid_string) do
```

Comment:
no need to convert UUID at all. It's a string. It's better to move this code to the validations module

Reply:
I would keep it, because if you have upper / mixed case string you can convert it to downcase once

Reply:
But it's neither a conversion nor parsing. You convert a string into a string. And we already have a function for such normalization: `String.downcase/1`. Ecto had to cast UUID because it's a real conversion from the string to binary representation for postgres.